File: omnigent/server/routes/dictation.py
"""Streaming dictation route: the transcription WebSocket.

This module hosts the server-side speech-to-text surface behind the
composer mic button (``designs/server-dictation.md``):

- ``WS /v1/dictation/stream`` — one connection per dictation take.

Availability is advertised as ``dictation_available`` on ``GET /v1/info``
(the web UI's boot-time capability probe); there is no separate probe
endpoint.

Wire protocol on the WebSocket
------------------------------

- **Client → server, binary frames**: raw 16 kHz mono s16le PCM. The
  browser worklet downsamples from the capture rate before sending.
- **Client → server, text frames**: JSON control messages.
  ``{"type": "stop"}`` asks the server to flush trailing audio and
  finish the take. Unknown shapes are ignored so future control
  messages don't break older servers.
- **Server → client, text frames**: JSON events.
    - ``{"type": "ready"}`` — sent once after the engine is ready;
      the client may start streaming audio.
    - ``{"type": "partial", "text": ...}`` — revisable in-progress
      utterance, throttled server-side.
    - ``{"type": "final", "text": ...}`` — an utterance completed by
      endpoint detection (a pause). The client appends it and clears
      its partial region.
    - ``{"type": "stopped", "text": ...}`` — reply to ``stop``: the
      flushed tail utterance (possibly empty). The server closes the
      socket after sending it.
    - ``{"type": "error", "message": ...}`` — fatal; the server closes.

Auth
----

Dictation is not session-scoped — the new-chat composer dictates before
any session exists — so the check is identity-level only, matching
``GET /v1/harnesses``: when an auth provider is configured the caller
must be authenticated (the WebSocket handshake carries identity via the
ingress/dev proxy exactly like the terminal-attach socket); in
single-user/dev mode the route is open.

Capacity
--------

Decoding is CPU-bound, so concurrent takes are capped (default 2,
``OMNIGENT_DICTATION_MAX_STREAMS``). Over-cap connections are accepted
and immediately closed with code 1013 (try again later) so the client
can distinguish "busy" from "broken".
"""

from __future__ import annotations

import asyncio
import contextlib
import json
import logging
import time
from collections.abc import Callable
from typing import Final

import anyio
from fastapi import APIRouter, WebSocket, WebSocketDisconnect, WebSocketException
from starlette import status

from omnigent.debug_logging import debug_event
from omnigent.server.auth import AuthProvider
from omnigent.server.dictation import (
    DictationEngine,
    DictationStreamHandle,
    get_engine,
    max_streams,
)

_logger = logging.getLogger(__name__)

_WS_CLOSE_TRY_AGAIN_LATER: Final[int] = 1013
_WS_CLOSE_INTERNAL_ERROR: Final[int] = 1011

#: Minimum interval between partial-transcript pushes. Keeps the socket
#: chatty enough for live text without a frame per audio chunk.
_PARTIAL_INTERVAL_S: Final[float] = 0.15


def create_dictation_router(
    *,
    auth_provider: AuthProvider | None = None,
    engine_provider: Callable[[], DictationEngine] | None = None,
) -> APIRouter:
    """Build the router carrying the dictation stream route.

    Wired into the FastAPI app under the ``/v1`` prefix in
    :func:`omnigent.server.app.create_app`.

    :param auth_provider: Optional provider used to authenticate the
        WebSocket handshake. ``None`` preserves single-user/dev
        behavior (open).
    :param engine_provider: Engine factory override for tests. Defaults
        to :func:`omnigent.server.dictation.get_engine`, which resolves
        the configured engine and loads models on first use.
    :returns: An :class:`APIRouter` carrying the stream route.
    """
    router = APIRouter()
    resolve_engine = engine_provider or get_engine
    # Router-scoped so each app (and each test app) gets its own cap.
    slots = asyncio.Semaphore(max_streams())

    @router.websocket("/dictation/stream")
    async def dictation_stream(websocket: WebSocket) -> None:
        """Transcribe one dictation take (see module docstring)."""
        if auth_provider is not None and auth_provider.get_user_id(websocket) is None:
            raise WebSocketException(
                code=status.WS_1008_POLICY_VIOLATION,
                reason="authentication required",
            )
        await websocket.accept()
        _logger.info(
            "dictation stream connected",
            extra=debug_event("dictation_stream", phase="connected"),
        )

        if slots.locked():
            await websocket.close(
                code=_WS_CLOSE_TRY_AGAIN_LATER,
                reason="dictation is at capacity; try again shortly",
            )
            return

        async with slots:
            # Engine construction loads model weights — seconds on first
            # use. Run it off-loop; later takes reuse the shared engine.
            try:
                engine = await asyncio.to_thread(resolve_engine)
                handle: DictationStreamHandle = await asyncio.to_thread(engine.create_stream)
            except Exception:
                _logger.exception(
                    "dictation engine failed to initialize",
                    extra=debug_event("dictation_stream", phase="error", stage="engine_init"),
                )
                with contextlib.suppress(RuntimeError):
                    await websocket.send_text(
                        json.dumps({"type": "error", "message": "dictation engine unavailable"})
                    )
                    await websocket.close(code=_WS_CLOSE_INTERNAL_ERROR)
                return
            # Release the take on every exit — normal stop, abrupt browser
            # disconnect, or a crash mid-send. For the in-process engines
            # close() just frees the recognizer stream, so a best-effort
            # close on the way out is enough.
            try:
                await websocket.send_text(json.dumps({"type": "ready"}))
                await _pump_dictation(websocket, handle)
            finally:
                _logger.info(
                    "dictation stream disconnected",
                    extra=debug_event("dictation_stream", phase="disconnected"),
                )
                # An abrupt disconnect tears the ASGI task down via
                # cancellation, which would cancel this close mid-await and
                # leak the take (the remote engine holds a worker slot until
                # close). Shield it so cleanup always completes.
                with anyio.CancelScope(shield=True):
                    try:
                        await asyncio.to_thread(handle.close)
                    except Exception:  # noqa: BLE001 - fall back to a direct close
                        # During teardown the loop's thread-pool executor may
                        # already be shutting down, so offloading raises rather
                        # than running close() — which would leak the take.
                        # close() is a quick, non-blocking free for every
                        # engine, so fall back to a direct call on the loop.
                        with contextlib.suppress(Exception):
                            handle.close()

    return router


async def _pump_dictation(websocket: WebSocket, handle: DictationStreamHandle) -> None:
    """Shuttle audio in and transcript events out until stop/disconnect.

    :param websocket: The accepted browser-facing WebSocket.
    :param handle: The per-connection recognizer stream.
    """
    last_partial_sent = ""
    last_partial_at = 0.0
    try:
        while True:
            message = await websocket.receive()
            if message.get("type") == "websocket.disconnect":
                return

            data = message.get("bytes")
            if data is not None:
                update = await asyncio.to_thread(handle.feed_pcm16, data)
                if update.finalized:
                    await websocket.send_text(
                        json.dumps({"type": "final", "text": update.finalized})
                    )
                    last_partial_sent = ""
                    last_partial_at = 0.0
                now = time.monotonic()
                if (
                    update.partial != last_partial_sent
                    and now - last_partial_at >= _PARTIAL_INTERVAL_S
                ):
                    await websocket.send_text(
                        json.dumps({"type": "partial", "text": update.partial})
                    )
                    last_partial_sent = update.partial
                    last_partial_at = now
                continue

            text_frame = message.get("text")
            if text_frame is None:
                continue
            try:
                control = json.loads(text_frame)
            except ValueError:
                continue
            if isinstance(control, dict) and control.get("type") == "stop":
                tail = await asyncio.to_thread(handle.finish)
                await websocket.send_text(json.dumps({"type": "stopped", "text": tail}))
                await websocket.close()
                return
            # Unknown control messages are ignored for forward compat.
    except WebSocketDisconnect:
        return
    except Exception:
        _logger.exception(
            "dictation stream failed",
            extra=debug_event("dictation_stream", phase="error", stage="pump"),
        )
        with contextlib.suppress(RuntimeError):
            await websocket.send_text(json.dumps({"type": "error", "message": "dictation failed"}))
            await websocket.close(code=_WS_CLOSE_INTERNAL_ERROR)

File: web/src/lib/dictation.ts
// Server-side dictation transport: mic → 16 kHz PCM → WS /v1/dictation/stream.
//
// ComposerMicButton uses this as the fallback when the browser Web Speech API
// has no working backend (Electron, Firefox/Chromium — see
// web/electron/README.md). One DictationSession is one dictation take: it
// owns the microphone stream, an AudioWorklet that downsamples the capture
// rate to 16 kHz mono s16le, and the WebSocket that streams those frames to
// the server and receives transcript events back. The wire protocol is
// documented in omnigent/server/routes/dictation.py; availability is gated
// by the `dictation_available` capability from GET /v1/info.
//
// The WebSocket URL rides the host seam (`resolveWebSocketUrl`) exactly like
// the terminal-attach and session-updates sockets, so embed hosts and the
// Vite dev proxy keep working. Identity rides the ingress/dev proxy on the
// handshake, as with those sockets.

import { getOmnigentHostConfig, resolveWebSocketUrl } from "@/lib/host";
import { modalHostId } from "@/lib/sessionHost";

/**
 * Build the `ws(s)://` URL for the dictation stream.
 *
 * When a host fetcher is installed, append `?omnigent_slice_key=<modal host>` —
 * the same query-param seam the session-updates and terminal-attach sockets use
 * (a browser WS handshake can't set request headers). Dictation is NOT
 * host-scoped — any replica transcribes a take equally, so correctness is
 * unaffected either way. It's a LOAD-BALANCING lever: without a key, the default
 * route hashes a whole tenant onto one replica, so a busy tenant's mic takes all
 * pile onto a single replica — bad for a CPU-bound engine with a per-replica
 * concurrent-take cap ({@link DictationBusyError} / `OMNIGENT_DICTATION_MAX_STREAMS`).
 * {@link modalHostId} is per-user (the host backing most of THIS user's
 * sessions), so keying by it spreads a tenant's takes across many replicas —
 * finer-grained than per-tenant. Omitted on an unsharded server and until the
 * modal host resolves (→ no key → default route).
 */
function buildDictationUrl(): string {
  const path = "/v1/dictation/stream";
  if (!getOmnigentHostConfig().fetcher) return resolveWebSocketUrl(path);
  const sliceKey = modalHostId();
  return resolveWebSocketUrl(
    sliceKey ? `${path}?omnigent_slice_key=${encodeURIComponent(sliceKey)}` : path,
  );
}

/** A transcript event pushed by the server over the dictation stream. */
export type DictationEvent =
  | { type: "ready" }
  | { type: "partial"; text: string }
  | { type: "final"; text: string }
  | { type: "stopped"; text: string }
  | { type: "error"; message: string };

export interface DictationSessionEvents {
  /** Revisable in-progress utterance (server-throttled to ~6 Hz). */
  onPartial: (text: string) => void;
  /** An utterance completed by a pause; append it and clear the partial. */
  onFinal: (text: string) => void;
  /** Fatal error after start. The session has already cleaned itself up. */
  onError: (message: string) => void;
}

/**
 * The server was reachable but at its concurrent-take cap (WS close 1013).
 * Transient by definition — callers should message "busy, try again",
 * not "unavailable".
 */
export class DictationBusyError extends Error {}

/**
 * Parse one text frame from the dictation socket into a typed event.
 * Returns null for frames that don't match the protocol (ignored for
 * forward compatibility, mirroring the server's posture on control
 * messages it doesn't know).
 */
export function parseDictationEvent(raw: string): DictationEvent | null {
  let data: unknown;
  try {
    data = JSON.parse(raw);
  } catch {
    return null;
  }
  if (typeof data !== "object" || data === null) return null;
  const frame = data as { type?: unknown; text?: unknown; message?: unknown };
  switch (frame.type) {
    case "ready":
      return { type: "ready" };
    case "partial":
    case "final":
    case "stopped":
      return typeof frame.text === "string" ? { type: frame.type, text: frame.text } : null;
    case "error":
      return typeof frame.message === "string" ? { type: "error", message: frame.message } : null;
    default:
      return null;
  }
}

// Client budgets must exceed the server's own worst cases or takes fail
// spuriously right when they'd have succeeded:
// - ready: engine construction loads model weights on the first take, and
//   the remote-relay path allows the worker 30 s for its own cold load
//   (_REMOTE_READY_TIMEOUT_S in omnigent/server/dictation.py).
// - stop: the relay waits up to 10 s (_REMOTE_STOP_TIMEOUT_S) for the
//   worker to flush the tail; resolving earlier would drop the user's
//   last words even though they were transcribed moments later.
const READY_TIMEOUT_MS = 40_000;
const STOP_TIMEOUT_MS = 15_000;

// How long stop() waits for the worklet to post its final partial chunk
// before tearing the audio graph down. Message-port turnaround is
// milliseconds; this is only a stuck-worklet backstop.
const FLUSH_TIMEOUT_MS = 250;

/** WS close code the server sends when at its concurrent-take cap. */
const WS_CLOSE_TRY_AGAIN_LATER = 1013;

const TARGET_RATE = 16_000;

// AudioWorklet processor, inlined as a Blob module so no separate asset has
// to survive the Vite build. Linear-interpolation downsample from the
// context capture rate to 16 kHz, Float32 → Int16, posted in 100 ms chunks.
// (When the context already runs at 16 kHz — we ask for that — the step is
// 1 and the loop degenerates to a plain format conversion.)
//
// Any message on the port means "flush": the partially-filled chunk is
// posted, then a null marker — so stop() can capture trailing speech that
// hasn't crossed the 100 ms boundary before tearing the graph down.
const WORKLET_SOURCE = `
const TARGET_RATE = ${TARGET_RATE};
const CHUNK_SAMPLES = TARGET_RATE / 10; // 100 ms per posted chunk
class Pcm16Downsampler extends AudioWorkletProcessor {
  constructor() {
    super();
    this.step = sampleRate / TARGET_RATE;
    this.pos = 0; // fractional read position, carried across blocks
    this.pending = new Int16Array(CHUNK_SAMPLES);
    this.filled = 0;
    this.port.onmessage = () => {
      if (this.filled > 0) {
        const out = this.pending.slice(0, this.filled);
        this.filled = 0;
        this.port.postMessage(out, [out.buffer]);
      }
      this.port.postMessage(null);
    };
  }
  process(inputs) {
    const channel = inputs[0] && inputs[0][0];
    if (!channel || channel.length === 0) return true;
    let pos = this.pos;
    while (pos < channel.length) {
      const i = Math.floor(pos);
      const s0 = channel[i];
      const s1 = i + 1 < channel.length ? channel[i + 1] : s0;
      const sample = s0 + (s1 - s0) * (pos - i);
      const clamped = Math.max(-1, Math.min(1, sample));
      this.pending[this.filled++] = clamped < 0 ? clamped * 0x8000 : clamped * 0x7fff;
      if (this.filled === CHUNK_SAMPLES) {
        const out = this.pending;
        this.pending = new Int16Array(CHUNK_SAMPLES);
        this.filled = 0;
        this.port.postMessage(out, [out.buffer]);
      }
      pos += this.step;
    }
    this.pos = pos - channel.length;
    return true;
  }
}
registerProcessor("omnigent-pcm16-downsampler", Pcm16Downsampler);
`;

let cachedWorkletUrl: string | null = null;

function workletUrl(): string {
  if (cachedWorkletUrl === null) {
    cachedWorkletUrl = URL.createObjectURL(
      new Blob([WORKLET_SOURCE], { type: "application/javascript" }),
    );
  }
  return cachedWorkletUrl;
}

/**
 * One live dictation take against the server recognizer.
 *
 * Construct via {@link DictationSession.start}, which resolves once the
 * mic, audio graph, and socket handshake are all up — so a resolved
 * session is guaranteed to be streaming. End it with {@link stop} (flushes
 * the tail utterance) or {@link cancel} (immediate teardown, e.g. unmount).
 */
export class DictationSession {
  private readonly events: DictationSessionEvents;
  private readonly ws: WebSocket;
  private readonly mediaStream: MediaStream;
  private readonly audioContext: AudioContext;
  private readonly workletNode: AudioWorkletNode;
  private stopResolve: ((tail: string) => void) | null = null;
  private flushResolve: (() => void) | null = null;
  private closed = false;

  private constructor(
    events: DictationSessionEvents,
    ws: WebSocket,
    mediaStream: MediaStream,
    audioContext: AudioContext,
    workletNode: AudioWorkletNode,
  ) {
    this.events = events;
    this.ws = ws;
    this.mediaStream = mediaStream;
    this.audioContext = audioContext;
    this.workletNode = workletNode;

    workletNode.port.onmessage = (msg: MessageEvent<Int16Array<ArrayBuffer> | null>) => {
      if (msg.data === null) {
        // Flush marker: the worklet has posted everything it had.
        this.flushResolve?.();
        this.flushResolve = null;
        return;
      }
      if (ws.readyState === WebSocket.OPEN) ws.send(msg.data.buffer);
    };
    ws.onmessage = (msg) => {
      if (typeof msg.data !== "string") return;
      const event = parseDictationEvent(msg.data);
      if (event === null) return;
      if (event.type === "partial") this.events.onPartial(event.text);
      else if (event.type === "final") this.events.onFinal(event.text);
      else if (event.type === "stopped") this.resolveStop(event.text);
      else if (event.type === "error") this.fail(event.message);
    };
    ws.onclose = () => {
      // A close during stop() is the normal end of a take; any other
      // close means the server went away mid-dictation.
      if (this.stopResolve !== null) this.resolveStop("");
      else if (!this.closed) this.fail("Dictation connection closed");
    };
  }

  /**
   * Acquire the mic, open the socket, and wait for the server's ready
   * handshake. Rejects (with everything torn down) when the mic is
   * denied, the socket fails, the server is at capacity
   * ({@link DictationBusyError}), or the engine never comes up.
   */
  static async start(events: DictationSessionEvents): Promise<DictationSession> {
    const mediaStream = await navigator.mediaDevices.getUserMedia({
      audio: { channelCount: 1, echoCancellation: true, noiseSuppression: true },
    });

    let ws: WebSocket | null = null;
    let audioContext: AudioContext | null = null;
    try {
      ws = new WebSocket(buildDictationUrl());
      ws.binaryType = "arraybuffer";
      await waitForReady(ws);
      // Detect a close during the async audio-graph setup below: the
      // handler-swap in the constructor would otherwise never see it and
      // start() would resolve a dead session that silently drops audio.
      let closedDuringSetup = false;
      const markClosed = () => {
        closedDuringSetup = true;
      };
      ws.addEventListener("close", markClosed);

      // Ask for the target rate directly — Chrome/Firefox resample the
      // capture for us and the worklet's downsampler becomes a no-op.
      // Some platforms reject the hint; the worklet handles any rate.
      try {
        audioContext = new AudioContext({ sampleRate: TARGET_RATE });
      } catch {
        audioContext = new AudioContext();
      }
      await audioContext.audioWorklet.addModule(workletUrl());
      const source = audioContext.createMediaStreamSource(mediaStream);
      const node = new AudioWorkletNode(audioContext, "omnigent-pcm16-downsampler");
      // The worklet only renders while it reaches the destination; route
      // it through a muted gain so nothing is audible.
      const mute = audioContext.createGain();
      mute.gain.value = 0;
      source.connect(node);
      node.connect(mute);
      mute.connect(audioContext.destination);

      ws.removeEventListener("close", markClosed);
      if (closedDuringSetup || ws.readyState !== WebSocket.OPEN) {
        throw new Error("dictation connection closed during setup");
      }
      return new DictationSession(events, ws, mediaStream, audioContext, node);
    } catch (error) {
      for (const track of mediaStream.getTracks()) track.stop();
      if (audioContext && audioContext.state !== "closed") void audioContext.close();
      ws?.close();
      throw error;
    }
  }

  /**
   * End the take: flush the worklet's trailing samples (so the last words
   * make it to the recognizer), release the mic, ask the server to flush,
   * and resolve with the flushed tail utterance ("" on timeout/close).
   */
  async stop(): Promise<string> {
    if (this.closed) return "";
    await this.flushWorklet();
    this.teardownAudio();
    if (this.ws.readyState !== WebSocket.OPEN) {
      this.closed = true;
      return "";
    }
    return new Promise<string>((resolve) => {
      this.stopResolve = resolve;
      this.ws.send(JSON.stringify({ type: "stop" }));
      setTimeout(() => this.resolveStop(""), STOP_TIMEOUT_MS);
    });
  }

  /** Immediate teardown without waiting for the tail (unmount, disable). */
  cancel(): void {
    this.teardownAudio();
    this.closed = true;
    this.ws.close();
  }

  /** Ask the worklet to post its partial chunk; wait for its marker. */
  private flushWorklet(): Promise<void> {
    return new Promise<void>((resolve) => {
      this.flushResolve = resolve;
      try {
        this.workletNode.port.postMessage("flush");
      } catch {
        resolve();
        return;
      }
      setTimeout(resolve, FLUSH_TIMEOUT_MS);
    });
  }

  private resolveStop(tail: string): void {
    this.closed = true;
    const resolve = this.stopResolve;
    this.stopResolve = null;
    this.ws.close();
    resolve?.(tail);
  }

  private fail(message: string): void {
    this.closed = true;
    this.teardownAudio();
    this.ws.close();
    this.events.onError(message);
  }

  private teardownAudio(): void {
    for (const track of this.mediaStream.getTracks()) track.stop();
    if (this.audioContext.state !== "closed") void this.audioContext.close();
  }
}

/**
 * Resolve when the server sends its ready frame; reject on error frame,
 * close (typed {@link DictationBusyError} for the 1013 at-capacity close),
 * or timeout.
 */
function waitForReady(ws: WebSocket): Promise<void> {
  return new Promise((resolve, reject) => {
    const timer = setTimeout(() => {
      reject(new Error("dictation server did not become ready"));
    }, READY_TIMEOUT_MS);
    ws.onmessage = (msg) => {
      if (typeof msg.data !== "string") return;
      const event = parseDictationEvent(msg.data);
      if (event?.type === "ready") {
        clearTimeout(timer);
        resolve();
      } else if (event?.type === "error") {
        // The engine failed to initialize; surface its message rather
        // than the generic close that follows.
        clearTimeout(timer);
        reject(new Error(event.message));
      }
    };
    ws.onerror = () => {
      clearTimeout(timer);
      reject(new Error("dictation connection failed"));
    };
    ws.onclose = (event) => {
      clearTimeout(timer);
      reject(
        event.code === WS_CLOSE_TRY_AGAIN_LATER
          ? new DictationBusyError("dictation is at capacity")
          : new Error("dictation connection closed"),
      );
    };
  });
}
